synapse/handlers/space_summary.py

```diff
@@ -408,6 +521,7 @@ async def _build_room_entry(self, room_id: str) -> JsonDict:
             "guest_can_join": stats["guest_access"] == "can_join",
             "creation_ts": create_event.origin_server_ts,
             "room_type": room_type,
+            "allowed_spaces": allowed_spaces if allowed_spaces else None,
```

Comment:
I wonder if we should only be returning the list of spaces that the origin is in? I don't know to what extent we care about leaking the existence of spaces?

Reply:
Oh, interesting. The issue with that is we don't necessarily know if the origin is in them (if we're not in them ourselves), so I think that would make the edge-case at the bottom of https://github.com/matrix-org/matrix-doc/blob/clokep/restricted-rooms/proposals/3083-restricted-rooms.md#discovery-of-restricted-rooms worse:

> Note that there are still potential situations where each server individually doesn't have enough information to properly return the full summary, but these do not seem reasonable in what is considered a normal structure of spaces. (E.g. in the above example, if the remote server is not in the space and does not know whether the server is in the space or not it cannot return the room.)

So if we further filter this list down we might be making *more* situations where the servers cannot agree who is in the room. I'll need to think about it though...